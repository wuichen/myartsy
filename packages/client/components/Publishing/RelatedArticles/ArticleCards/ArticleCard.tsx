import { Box, color, Flex, Sans, Serif } from "@artsy/palette"
import { compact } from "lodash"
import React, { Component } from "react"
import track, { TrackingProp } from "react-tracking"
import styled from "styled-components"

import { media } from "Components/Helpers"
import { Byline } from "Components/Publishing/Byline/Byline"
import { Date as DateComponent } from "Components/Publishing/Byline/Date"
import { formatTime, getMediaDate } from "Components/Publishing/Constants"
import { IconVideoPlay } from "Components/Publishing/Icon/IconVideoPlay"
import { crop } from "Utils/resizer"

interface Props {
  article?: any
  color?: string
  editing?: boolean
  editDate?: any
  editDescription?: any
  editTitle?: any
  editImage?: any
  series?: any
  tracking?: TrackingProp
}

@track()
export class ArticleCard extends Component<Props> {
  public static defaultProps: Partial<Props>

  isUnpublishedMedia = () => {
    const { article } = this.props
    return (
      article.layout === "video" && article.media && !article.media.published
    )
  }

  isEditing = () => {
    const {
      editDate,
      editDescription,
      editImage,
      editTitle,
      editing,
    } = this.props

    return editing || editDate || editDescription || editImage || editTitle
  }

  renderDate = () => {
    const { article, editDate } = this.props
    const hasMedia =
      article.media && compact(Object.values(article.media)).length

    if (editDate) {
      return (
        <Sans size="3t" weight="medium">
          {editDate}
        </Sans>
      )
    } else if (hasMedia) {
      return this.renderMediaDate()
    } else {
      return (
        <Byline
          article={article}
          color={this.props.color}
          size="3t"
          layout="condensed"
        />
      )
    }
  }

  renderMediaDate = () => {
    const { article } = this.props
    const mediaDate = getMediaDate(article)
    const date = article.layout === "video" ? mediaDate : article.published_at
    const formatedDate = new Date(date).toISOString()

    if (this.isUnpublishedMedia()) {
      return (
        <Sans size="3t" weight="medium">
          <Flex alignItems="flex-end">
            <Box mr="5px">Available</Box>
            <DateComponent format="monthYear" date={mediaDate} />
          </Flex>
        </Sans>
      )
    } else {
      return <DateComponent layout="condensed" size="3t" date={formatedDate} />
    }
  }

  renderMediaCoverInfo = () => {
    const { article } = this.props

    if (this.isUnpublishedMedia()) {
      return (
        <MediaContainer>
          <Sans size={["8", "8", "10", "10"]}>Coming Soon</Sans>
        </MediaContainer>
      )
    } else {
      return (
        <MediaContainer justifyContent="space-between" alignItems="flex-end">
          <IconVideoPlay color={this.props.color} />
          <Sans size="3t" weight="medium" lineHeight="1em">
            {formatTime(article.media.duration)}
          </Sans>
        </MediaContainer>
      )
    }
  }

  openLink = e => {
    e.preventDefault()

    if (!this.isUnpublishedMedia() && !this.isEditing()) {
      window.open(e.currentTarget.attributes.href.value)
    }

    this.props.tracking.trackEvent({
      action: "Click",
      label: "Related article card",
    })
  }

  render() {
    const {
      article,
      editDescription,
      editImage,
      editTitle,
      series,
    } = this.props
    const { layout } = article
    const isUnpublishedMedia = this.isUnpublishedMedia()

    const description = editDescription ? editDescription : article.description

    const title = editTitle
      ? editTitle
      : article.thumbnail_title || article.title

    return (
      <ArticleCardContainer
        href={isUnpublishedMedia ? "" : article.slug}
        color={this.props.color}
        className="ArticleCard"
        published={!isUnpublishedMedia}
        onClick={this.openLink}
      >
        <Flex
          flexDirection={["column-reverse", "column-reverse", "row", "row"]}
          p={[20, 20, 30, 30]}
        >
          <Flex
            width={["100%", "100%", "50%", "50%"]}
            mb={[0, 0, "5px", "5px"]}
            flexDirection="column"
            justifyContent="space-between"
          >
            <div>
              {series && (
                <Sans size="3t" weight="medium" mb={1}>
                  {series.title}
                </Sans>
              )}
              <Sans size={["8", "8", "10", "10"]} mb={20}>
                {title}
              </Sans>

              <Serif size={["4", "4", "5", "5"]} mb={20}>
                {description}
              </Serif>
            </div>
            {this.renderDate()}
          </Flex>

          <ArticleCardImageContainer
            width={["100%", "100%", "50%", "50%"]}
            ml={[0, 0, 30, 30]}
            mb={[10, 10, 0, 0]}
          >
            {editImage ? (
              editImage
            ) : (
              <Image
                src={crop(article.thumbnail_image, { width: 680, height: 450 })}
              />
            )}
            {article.media && layout === "video" && this.renderMediaCoverInfo()}
          </ArticleCardImageContainer>
        </Flex>
      </ArticleCardContainer>
    )
  }
}

ArticleCard.defaultProps = {
  color: color("black100"),
}

const Image = styled.img`
  width: 100%;
  height: 100%;
  object-fit: cover;
  object-position: center;
  transition: opacity 0.15s;
  display: block;
`

export const ArticleCardImageContainer = styled(Box)`
  position: relative;
  height: inherit;
  background: white;
`

export const ArticleCardContainer = styled.a<{ published: boolean }>`
  border: 1px solid;
  border-radius: 2px;
  color: ${props => props.color};
  cursor: ${props => (props.published ? "pointer" : "default")};
  text-decoration: none;
  display: block;

  ${Image} {
    opacity: ${props => (props.published ? "1" : "0.7")};
  }

  &:hover {
    ${Image} {
      opacity: 0.7;
    }
  }
`

const MediaContainer = styled(Flex)`
  position: absolute;
  left: 20px;
  right: 20px;
  bottom: 20px;

  svg {
    height: 40px;
  }
  ${media.md`
    svg {
      height: 30px;
    }
  `};
`
