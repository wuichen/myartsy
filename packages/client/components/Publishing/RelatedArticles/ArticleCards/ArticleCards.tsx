import { Box } from "@artsy/palette"
import React from "react"
import styled from "styled-components"

import { ArticleData } from "Components/Publishing/Typings"
import { ArticleCard } from "./ArticleCard"

interface Props {
  series?: ArticleData
  color?: string
  relatedArticles?: any
}

export const ArticleCards: React.SFC<Props> = props => {
  const { color, relatedArticles, series } = props

  return (
    <Box color={color}>
      {relatedArticles.map((relatedArticle, i) => {
        return (
          <ArticlesWrapper mb={[40, 60]} key={i}>
            <ArticleCard
              article={relatedArticle}
              color={color}
              series={series}
            />
          </ArticlesWrapper>
        )
      })}
    </Box>
  )
}

// Used to target wrapper in other components
export const ArticlesWrapper = styled(Box)``
