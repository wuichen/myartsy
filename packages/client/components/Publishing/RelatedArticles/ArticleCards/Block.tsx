import { Box, color } from "@artsy/palette"
import React from "react"
import styled from "styled-components"

import { unica } from "Assets/Fonts"
import {
  Vertical,
  VerticalOrSeriesTitle,
} from "Components/Publishing/Sections/VerticalOrSeriesTitle"
import { SeriesAbout } from "Components/Publishing/Series/SeriesAbout"
import { ArticleData } from "Components/Publishing/Typings"
import { ArticleCards } from "./ArticleCards"

interface Props {
  article?: ArticleData
  color?: string
  relatedArticles?: any
}

export const ArticleCardsBlock: React.SFC<Props> = props => {
  const { article, relatedArticles } = props
  const { seriesArticle } = article

  return (
    <ArticleCardsContainer color={props.color}>
      {(relatedArticles || article.relatedArticles) && (
        <Box maxWidth={1200} mx="auto">
          <VerticalOrSeriesTitle
            article={article}
            color={props.color}
            prefix="More in "
          />
          <ArticleCards
            relatedArticles={relatedArticles || article.relatedArticles}
            series={seriesArticle}
            color={props.color}
          />
        </Box>
      )}
      {seriesArticle && (
        <Box maxWidth={1200} mx="auto" pt={[40, 40, 60]} pb={100}>
          <SeriesAbout article={seriesArticle} color={props.color} />
        </Box>
      )}
    </ArticleCardsContainer>
  )
}

ArticleCardsBlock.defaultProps = {
  color: color("black100"),
}

export const ArticleCardsContainer = styled.div`
  color: ${props => props.color};

  ${Vertical} {
    ${unica("s32")};
    width: 100%;
    margin-bottom: 40px;

    a {
      border-bottom: 2px solid;
    }
  }
`
