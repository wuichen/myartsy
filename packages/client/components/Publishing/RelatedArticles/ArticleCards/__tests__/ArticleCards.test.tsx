import {
  SeriesArticleSponsored,
  StandardArticle,
  VideoArticle,
} from "Components/Publishing/Fixtures/Articles"
import { mount } from "enzyme"
import "jest-styled-components"
import React from "react"
import renderer from "react-test-renderer"
import { ArticleCard } from "../ArticleCard"
import { ArticleCards } from "../ArticleCards"

describe("ArticleCard", () => {
  it("renders properly", () => {
    const component = renderer
      .create(
        <ArticleCards relatedArticles={[StandardArticle, VideoArticle]} />
      )
      .toJSON()

    expect(component).toMatchSnapshot()
  })

  it("renders a list of articles properly", () => {
    const component = mount(
      <ArticleCards relatedArticles={[StandardArticle, VideoArticle]} />
    )
    expect(component.find(ArticleCard).length).toBe(2)
  })

  it("renders a list of articles in a series properly", () => {
    const component = mount(
      <ArticleCards
        series={SeriesArticleSponsored}
        relatedArticles={[StandardArticle, VideoArticle]}
      />
    )
    expect(component.find(ArticleCard).length).toBe(2)
  })
})
