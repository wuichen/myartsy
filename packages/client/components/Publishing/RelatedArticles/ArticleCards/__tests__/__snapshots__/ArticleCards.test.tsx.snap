// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleCard renders properly 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column-reverse;
  -ms-flex-direction: column-reverse;
  flex-direction: column-reverse;
  padding: 20px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: 0px;
  width: 100%;
}

.c4 {
  font-family: Unica77LLWebRegular,'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 28px;
  line-height: 36px;
  margin-bottom: 20px;
}

.c5 {
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 18px;
  line-height: 26px;
  margin-bottom: 20px;
}

.c7 {
  font-family: Unica77LLWebMedium,'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-weight: 500;
  font-size: 14px;
  line-height: 20px;
}

.c15 {
  font-family: Unica77LLWebMedium,'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-weight: 500;
  font-size: 14px;
  line-height: 1em;
}

.c8 {
  margin: 10px 20px 0 0;
}

.c8::before {
  content: "";
  display: inline-block;
  width: 8px;
  height: 8px;
  border-radius: 50%;
  margin: 0 5px 1px 0;
  background-color: #000;
}

.c9 {
  margin: 5px 20px 0 0;
  white-space: nowrap;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  color: #000;
}

.c14 {
  width: 32px;
  height: 32px;
}

.c12 {
  width: 100%;
  height: 100%;
  object-fit: cover;
  object-position: center;
  -webkit-transition: opacity 0.15s;
  transition: opacity 0.15s;
  display: block;
}

.c10 {
  margin-bottom: 10px;
  margin-left: 0px;
  width: 100%;
  position: relative;
  height: inherit;
  background: white;
}

.c1 {
  border: 1px solid;
  border-radius: 2px;
  color: #000;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: block;
}

.c1 .c11 {
  opacity: 1;
}

.c1:hover .c11 {
  opacity: 0.7;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  position: absolute;
  left: 20px;
  right: 20px;
  bottom: 20px;
}

.c13 svg {
  height: 40px;
}

.c0 {
  margin-bottom: 40px;
}

@media screen and (min-width:40em) {
  .c2 {
    -webkit-flex-direction: column-reverse;
    -ms-flex-direction: column-reverse;
    flex-direction: column-reverse;
  }
}

@media screen and (min-width:52em) {
  .c2 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media screen and (min-width:64em) {
  .c2 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media screen and (min-width:40em) {
  .c2 {
    padding: 20px;
  }
}

@media screen and (min-width:52em) {
  .c2 {
    padding: 30px;
  }
}

@media screen and (min-width:64em) {
  .c2 {
    padding: 30px;
  }
}

@media screen and (min-width:40em) {
  .c3 {
    margin-bottom: 0px;
  }
}

@media screen and (min-width:52em) {
  .c3 {
    margin-bottom: 5px;
  }
}

@media screen and (min-width:64em) {
  .c3 {
    margin-bottom: 5px;
  }
}

@media screen and (min-width:40em) {
  .c3 {
    width: 100%;
  }
}

@media screen and (min-width:52em) {
  .c3 {
    width: 50%;
  }
}

@media screen and (min-width:64em) {
  .c3 {
    width: 50%;
  }
}

@media screen and (min-width:40em) {
  .c4 {
    font-size: 28px;
  }
}

@media screen and (min-width:52em) {
  .c4 {
    font-size: 42px;
  }
}

@media screen and (min-width:64em) {
  .c4 {
    font-size: 42px;
  }
}

@media screen and (min-width:40em) {
  .c4 {
    line-height: 36px;
  }
}

@media screen and (min-width:52em) {
  .c4 {
    line-height: 50px;
  }
}

@media screen and (min-width:64em) {
  .c4 {
    line-height: 50px;
  }
}

@media screen and (min-width:40em) {
  .c5 {
    font-size: 18px;
  }
}

@media screen and (min-width:52em) {
  .c5 {
    font-size: 22px;
  }
}

@media screen and (min-width:64em) {
  .c5 {
    font-size: 22px;
  }
}

@media screen and (min-width:40em) {
  .c5 {
    line-height: 26px;
  }
}

@media screen and (min-width:52em) {
  .c5 {
    line-height: 32px;
  }
}

@media screen and (min-width:64em) {
  .c5 {
    line-height: 32px;
  }
}

@media screen and (min-width:40em) {
  .c10 {
    margin-bottom: 10px;
    margin-left: 0px;
  }
}

@media screen and (min-width:52em) {
  .c10 {
    margin-bottom: 0px;
    margin-left: 30px;
  }
}

@media screen and (min-width:64em) {
  .c10 {
    margin-bottom: 0px;
    margin-left: 30px;
  }
}

@media screen and (min-width:40em) {
  .c10 {
    width: 100%;
  }
}

@media screen and (min-width:52em) {
  .c10 {
    width: 50%;
  }
}

@media screen and (min-width:64em) {
  .c10 {
    width: 50%;
  }
}

@media (max-width:900px) {
  .c13 svg {
    height: 30px;
  }
}

@media screen and (min-width:40em) {
  .c0 {
    margin-bottom: 60px;
  }
}

<div
  className=""
>
  <div
    className="c0"
  >
    <a
      className="ArticleCard c1"
      color="#000"
      href="new-yorks-next-art-district"
      onClick={[Function]}
    >
      <div
        className="c2"
      >
        <div
          className="c3"
          width={
            Array [
              "100%",
              "100%",
              "50%",
              "50%",
            ]
          }
        >
          <div>
            <div
              className="c4"
              fontFamily="Unica77LLWebRegular, 'Helvetica Neue', Helvetica, Arial, sans-serif"
              fontSize={
                Array [
                  "28px",
                  "28px",
                  "42px",
                  "42px",
                ]
              }
            >
              New York's Next Art District
            </div>
            <div
              className="c5"
              fontFamily="'Adobe Garamond W08', 'adobe-garamond-pro', 'AGaramondPro-Regular', 'Times New Roman', Times, serif"
              fontSize={
                Array [
                  "18px",
                  "18px",
                  "22px",
                  "22px",
                ]
              }
            >
              Land exhibitions, make influential contacts, and gain valuable feedback about your work.
            </div>
          </div>
          <div
            className="Byline c6"
            color="#000"
          >
            <div
              className="c7"
              fontFamily={
                Object {
                  "fontFamily": "Unica77LLWebMedium, 'Helvetica Neue', Helvetica, Arial, sans-serif",
                  "fontWeight": 500,
                }
              }
              fontSize="14px"
            >
              <div
                className="c8"
                color="#000"
              >
                Casey Lesser
              </div>
            </div>
            <div
              className="c9"
            >
              <div
                className="c7"
                fontFamily={
                  Object {
                    "fontFamily": "Unica77LLWebMedium, 'Helvetica Neue', Helvetica, Arial, sans-serif",
                    "fontWeight": 500,
                  }
                }
                fontSize="14px"
              >
                May 19, 2017
              </div>
            </div>
          </div>
        </div>
        <div
          className="c10"
          width={
            Array [
              "100%",
              "100%",
              "50%",
              "50%",
            ]
          }
        >
          <img
            className="c11 c12"
            src="https://d7hftxdivxxvm.cloudfront.net?resize_to=fill&src=https%3A%2F%2Fartsy-media-uploads.s3.amazonaws.com%2F7lsxxsw0qPAuKl37jEYitw%252Farticle%2Basset%2B1-hig%2Bres%2Bcopy.jpg&width=680&height=450&quality=95"
          />
        </div>
      </div>
    </a>
  </div>
  <div
    className="c0"
  >
    <a
      className="ArticleCard c1"
      color="#000"
      href="joanne-artman-gallery-poetry-naturerefinement-form"
      onClick={[Function]}
    >
      <div
        className="c2"
      >
        <div
          className="c3"
          width={
            Array [
              "100%",
              "100%",
              "50%",
              "50%",
            ]
          }
        >
          <div>
            <div
              className="c4"
              fontFamily="Unica77LLWebRegular, 'Helvetica Neue', Helvetica, Arial, sans-serif"
              fontSize={
                Array [
                  "28px",
                  "28px",
                  "42px",
                  "42px",
                ]
              }
            >
              New Study Shows the Gender Pay Gap for Artists Is Not So Simple
            </div>
            <div
              className="c5"
              fontFamily="'Adobe Garamond W08', 'adobe-garamond-pro', 'AGaramondPro-Regular', 'Times New Roman', Times, serif"
              fontSize={
                Array [
                  "18px",
                  "18px",
                  "22px",
                  "22px",
                ]
              }
            >
              The elegant spiral of the Nautilus shell, the sinuous pattern of the banks of a river, or the swirling vortex street of clouds - patterns exist.
            </div>
          </div>
          <div
            className="c9"
          >
            <div
              className="c7"
              fontFamily={
                Object {
                  "fontFamily": "Unica77LLWebMedium, 'Helvetica Neue', Helvetica, Arial, sans-serif",
                  "fontWeight": 500,
                }
              }
              fontSize="14px"
            >
              Aug 28, 2017
            </div>
          </div>
        </div>
        <div
          className="c10"
          width={
            Array [
              "100%",
              "100%",
              "50%",
              "50%",
            ]
          }
        >
          <img
            className="c11 c12"
            src="https://d7hftxdivxxvm.cloudfront.net?resize_to=fill&src=https%3A%2F%2Fartsy-media-uploads.s3.amazonaws.com%2FwHFgQlrTrHav5O6bQRJ0dg%252FUntitled%2BSuspended_30x67x33%2B%25282%2529_sm%2Bcropped.jpg&width=680&height=450&quality=95"
          />
          <div
            className="c13"
          >
            <svg
              className="c14"
              version="1.1"
              viewBox="0 0 40 56"
              x="0px"
              xmlns="http://www.w3.org/2000/svg"
              y="0px"
            >
              <g
                fill="none"
                fillRule="evenodd"
                stroke="none"
                strokeWidth="1"
              >
                <polygon
                  fill="#000"
                  points="0 0 0 56 39.3600006 28"
                />
              </g>
            </svg>
            <div
              className="c15"
              fontFamily={
                Object {
                  "fontFamily": "Unica77LLWebMedium, 'Helvetica Neue', Helvetica, Arial, sans-serif",
                  "fontWeight": 500,
                }
              }
              fontSize="14px"
            >
              02:28
            </div>
          </div>
        </div>
      </div>
    </a>
  </div>
</div>
`;
